### PR TITLE
Add OS, Wine and Proton logging

### PR DESCRIPTION
**Pull request by Jan200101**
Relevant log on Fedora 38 running NorthstarProton 8.1
```
[15:16:48] [NORTHSTAR] [info] NorthstarLauncher version: 0.0.0.1+dev
[15:16:48] [NORTHSTAR] [info] Command line: "Z:\home\sentry\.local\share\Steam\steamapps\common\Titanfall2\Titanfall2.exe"   -northstar --use-gl=osmesa
[15:16:48] [NORTHSTAR] [info] Using profile: R2Northstar
[15:16:48] [NORTHSTAR] [info] Operating System: Linux (Wine)
[15:16:48] [NORTHSTAR] [info] Wine build: wine-8.1.r0.gdde73435 ( TkG Staging Esync Fsync )
[15:16:48] [NORTHSTAR] [info] Proton build: NorthstarProton-8.1-1
```

### CODE REVIEW

File: NorthstarDLL/logging/logging.cpp [hunk 1]
```diff
@@ -220,3 +222,94 @@ void NS::log::FlushLoggers()
...
 }
+
+// Wine specific functions
+typedef const char *(CDECL *wine_get_host_version_type)(const char**, const char**);
+wine_get_host_version_type wine_get_host_version;
+
+typedef const char *(CDECL *wine_get_build_id_type)(void);
+wine_get_build_id_type wine_get_build_id;
+
+typedef char *(CDECL *wine_get_unix_file_name_type)(LPCWSTR);
```
Comment (Jan200101) [inline NorthstarDLL/logging/logging.cpp:233]: These are leftovers from a previous attempt of this that need to be removed.
In a default Wine environment its possible to directly read files, from the host fs simply by translating the paths over.

Proton runs in a container with a static user `steamuser`, which means it cannot read user files unless specifically exposed by the container setup.

File: NorthstarDLL/logging/logging.cpp [hunk 2]
```diff
@@ -220,3 +222,94 @@ void NS::log::FlushLoggers()
...
+					win_ver = "8.1";
+					break;
+				default:
+					break;
+			}
+		}
+		else if (osvi.dwMajorVersion == 10)
+		{
+			win_ver = "10";
+		}
```
Comment (Jan200101) [inline NorthstarDLL/logging/logging.cpp:311]: The relevant table I was consulting did not have an entry for Windows 11.

I assume its just `dwMajorVersion == 11`, but I have no setup to test this (neither do I really want to install Windows 11)

File: NorthstarDLL/logging/logging.cpp [hunk 3]
```diff
@@ -220,3 +222,94 @@ void NS::log::FlushLoggers()
...
+	spdlog::info("Command line: {}", GetCommandLineA());
+	spdlog::info("Using profile: {}", GetNorthstarPrefix());
+
+	HMODULE ntdll = GetModuleHandleA("ntdll.dll");
+	if(!ntdll)
+	{
+		// How did we get here
+		spdlog::info("Operating System: Unknown");
+		return;
+	}
```
Comment (Jan200101) [inline NorthstarDLL/logging/logging.cpp:251]: This is practically pointless, but its better to sanity check.

File: NorthstarDLL/logging/logging.cpp [hunk 4]
```diff
@@ -220,3 +222,94 @@ void NS::log::FlushLoggers()
...
+    	wine_get_unix_file_name = (wine_get_unix_file_name_type)GetProcAddress(GetModuleHandleA("KERNEL32"), "wine_get_unix_file_name");
+    	wine_get_dos_file_name = (wine_get_dos_file_name_type)GetProcAddress(GetModuleHandleA("KERNEL32"), "wine_get_dos_file_name");
+
+		const char* sysname;
+		wine_get_host_version(&sysname, NULL);
+
+		spdlog::info("Operating System: {} (Wine)", sysname);
+		spdlog::info("Wine build: {}", wine_get_build_id());
+
+		char* compatToolPtr = std::getenv("STEAM_COMPAT_TOOL_PATHS");
```
Comment (Jan200101) [inline NorthstarDLL/logging/logging.cpp:267]: `GetEnvironmentVariable` would be more winapi oriented, but getenv doesn't require us to manually deal with stuff, plus its standard.

**Comment by Jan200101** [comment]
relevant issue #138

File: NorthstarDLL/logging/logging.cpp (hunk 4 above)
Comment (Jan200101) [inline NorthstarDLL/logging/logging.cpp:267]: Also regarding this env var: its the absolute host path to the compat tool folder e.g. `/home/user/.local/share/Steam/compatibilitytools.d/NorthstarProton8.1-1`

File: NorthstarDLL/logging/logging.cpp [hunk 5]
```diff
@@ -220,3 +222,94 @@ void NS::log::FlushLoggers()
...
+				default:
+					break;
+			}
+		}
+		else if (osvi.dwMajorVersion == 10)
+		{
+			win_ver = "10";
+		}
+
+		spdlog::info("Operating System: Windows {}", win_ver);
```
Comment (GeckoEidechse) [inline NorthstarDLL/logging/logging.cpp:313]: I would maybe even just stick `osvi.dwMajorVersion` and `osvi.dwMinorVersion` into the log IMO. The <Win10 percentage is pretty low AFAIK and only shrinking. If we really need to we can still look up what e.g. `NT6.3` corresponds. To me it kinda just feels like blowing up source code size without much payoff

(bike-shedding)

Comment (Jan200101) [reply NorthstarDLL/logging/logging.cpp:313]: yeah sure, though perhaps we should document a link to resource we can use to reference the version numbers.

**Comment by Jan200101** [comment]
Ignore me casually bumping my head while working Winapi shit out

**Review by F1F7Y — CHANGES_REQUESTED** [review]
Starting with windows 8.1 [`GetVersionEx`](https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getversionexa) behaves differently. This is on Windows 10 Home 21H2: 
image

The only good replacement i found is [`RtlGetVersion`](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-rtlgetversion?redirectedfrom=MSDN)

**Comment by Jan200101** [comment]
> Starting with windows 8.1 [`GetVersionEx`](https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getversionexa) behaves differently.
> This is on Windows 10 Home 21H2: image

God do I love Microsoft.
Apparently it now returns the relevant Version in relations to the programs Manifest, which we don't have.

Sooo.... I know that Source has an internhal helper function to identify the relevant OS Version, surely Respawn didn't gut it and replace it with their own hacky solution, right? :clueless:

Their method (found in engine.dll) does nonsense checks that at the end just return Windows 8.1.
Annoying

> The only good replacement i found is [`RtlGetVersion`](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-rtlgetversion?redirectedfrom=MSDN)

Yep, that seems to be the best alternative.
Since this is a one-time thing, going kernel space should have practically no impact at all.
Also, the one you linked appears to be the relevant function in Driver Space, here is the link to the winapi one [`RtlGetVersion`](https://learn.microsoft.com/en-us/windows/win32/devnotes/rtlgetversion)

**Comment by Jan200101** [comment]
Turns out, RtlGetVersion is pain in the ass to work with.
Implemented it similar to how the wine callbacks are done, as suggested by this SO Answer https://stackoverflow.com/a/48749703/6331425

Needs re-testing

File: NorthstarDLL/logging/logging.cpp [hunk 6]
```diff
@@ -220,3 +223,70 @@ void NS::log::FlushLoggers()
...
+	else
+	{
+		// We are real Windows (hopefully)
+		const char* win_ver = "Unknown";
+
+		RTL_OSVERSIONINFOW osvi;
+		osvi.dwOSVersionInfoSize = sizeof(osvi);
+
+		RtlGetVersion = (RtlGetVersion_type)GetProcAddress(ntdll, "RtlGetVersion");
+		if (RtlGetVersion && RtlGetVersion(&osvi))
```
Comment (F1F7Y) [inline NorthstarDLL/logging/logging.cpp:281]: `STATUS_SUCCESS` ( What `RtlGetVersion` returns ) is 0x0 ( [docs](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/596a1078-e883-4972-9bbc-49e60bebca55) ). This means this check will fail and fall back to the else block.

Modifying the check for the return value yielded this:
image

**Review by F1F7Y — CHANGES_REQUESTED** [review]

File: NorthstarDLL/logging/logging.cpp (hunk 6 above)
Comment (Jan200101) [reply NorthstarDLL/logging/logging.cpp]: Yep, I knew this but ended up fumbling it up along the way.
It appears to work as expected, so I guess thats solved.

**Review by F1F7Y — APPROVED** [review]
Code looks good, will have to go through it in my PR ( #504 ) once merged though.

Tested on Windows 10, works correctly.

**Review by GeckoEidechse — APPROVED** [review]
Works fine on Windows 10

```logs
[21:46:20] [NORTHSTAR] [info] NorthstarLauncher version: 0.0.0.1+dev
[21:46:20] [NORTHSTAR] [info] Command line: "C:\Program Files (x86)\Steam\steamapps\common\Titanfall2/NorthstarLauncher.exe" 
[21:46:20] [NORTHSTAR] [info] Using profile: R2Northstar
[21:46:20] [NORTHSTAR] [info] Operating System: Windows (NT10.0)
[21:46:21] [NORTHSTAR] [info] Enabling hook _GetCommandLineA
[21:46:21] [NORTHSTAR] [info] Enabling hook _LoadLibraryExA
...
````

**Comment by BimNoah** [comment]
Works fine on Gentoo with NorthstarProton 8.1.1

```
[21:57:37] [NORTHSTAR] [info] NorthstarLauncher version: 0.0.0.1+dev
[21:57:37] [NORTHSTAR] [info] Command line: "Z:\home\noah\.local\share\Steam\steamapps\common\Titanfall2\Titanfall2.exe"   -noborder -windowed -northstar --use-gl=osmesa
[21:57:37] [NORTHSTAR] [info] Using profile: R2Northstar
[21:57:37] [NORTHSTAR] [info] Operating System: Linux (Wine)
[21:57:37] [NORTHSTAR] [info] Wine build: wine-8.1.r0.gdde73435 ( TkG Staging Esync Fsync )
[21:57:37] [NORTHSTAR] [info] Proton build: NorthstarProton-8.1-1
[21:57:37] [NORTHSTAR] [info] Enabling hook _GetCommandLineA
```

File: NorthstarDLL/logging/logging.cpp (hunk 1 above)
Comment (GeckoEidechse) [reply NorthstarDLL/logging/logging.cpp]: I guess this got removed?

**Pull request merged**